### PR TITLE
Fix Async GridFS Download to Stream small buffer bug

### PR DESCRIPTION
**Pull request by rozza**
When using a smaller than chunksize buffer, need to check the internal buffer has been read before trying to get the next chunk.

JAVA-2294

### CODE REVIEW

**Comment by sfitts** [comment]
Thanks for the fix.  This is a total nit, but the check on the size of `dst` in `hasResultsToProcess` isn't really needed given that a zero size buffer will result in the first case in the if triggering.  If you did take that out then you wouldn't need to pass in `dst`.

File: driver-async/src/main/com/mongodb/async/client/gridfs/GridFSDownloadStreamImpl.java [hunk 1]
```diff
@@ -181,23 +181,26 @@ public void onResult(final List<Document> result, final Throwable t) {
     private void processResults(final int previousAmountRead, final ByteBuffer dst, final SingleResultCallback<Integer> callback) {
         try {
             int amountRead = previousAmountRead;
-            while (currentPosition < fileInfo.getLength() && dst.remaining() > 0 && !resultsQueue.isEmpty()) {
-                if (buffer == null || bufferOffset == buffer.length) {
+            int amountToCopy = dst.remaining();
+            while (currentPosition < fileInfo.getLength() && amountToCopy > 0) {
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/gridfs/GridFSDownloadStreamImpl.java]: This loop condition is confusing.  It checks that amountToCopy > 0, but amountToCopy is not updated after that amount is copied into the destination buffer on line 198.  Just to check, I added 

     amountToCopy = dst.remaining();

below line 201 and the tests still pass.  So maybe entering the loop one extra time is harmless, but it looks wrong.

Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/gridfs/GridFSDownloadStreamImpl.java]: Its updated on the next iteration of the loop, after the buffer is populated. As you've seen we could potentially shortcut the extra loop by reassigning after 201 as well.

Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/gridfs/GridFSDownloadStreamImpl.java]: Updated - looks better now.

Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/gridfs/GridFSDownloadStreamImpl.java]: Agreed

**Comment by jyemin** [comment]
LGTM

**Pull request closed without merging**